### ISSUE TITLE
Extra dependencies for package on private repo uninstalled when running `poetry install`.

### ISSUE THREAD

**Issue by iksteen** [python-poetry/poetry#3224]
- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) Poetry version.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option).

- **OS version and name**: Arch Linux
- **Poetry version**: 1.1.3
- **Link of a [Gist](https://gist.github.com/) with the contents of your pyproject.toml file**: https://gist.github.com/iksteen/bd2e570b8986f348e136038439d6be19

## Issue
After installing a package with extras from a private repository and running `poetry install` again, the extra dependencies are uninstalled. For easy reproducibility I've created a private repository that's publicly accessible and contains a single package (iksteen-aaa).

- Start out with a fairly empty pyproject.toml: https://gist.github.com/iksteen/83a3d5b55a8274bfbb434d72676fea70
- Now add the iksteen-aaa package with the example extra:
```
$ poetry add iksteen-aaa -E example
Using version ^0.1.0 for iksteen-aaa

Updating dependencies
Resolving dependencies... (1.6s)

Writing lock file

Package operations: 9 installs, 0 updates, 0 removals

  • Installing pycparser (2.20)
  • Installing cffi (1.14.3)
  • Installing pyasn1 (0.4.8)
  • Installing six (1.15.0)
  • Installing cryptography (3.1.1)
  • Installing ecdsa (0.14.1)
  • Installing rsa (4.6)
  • Installing python-jose (3.2.0)
  • Installing iksteen-aaa (0.1.0)
```
- Run `poetry install` again and see the extra dependencies getting uninstalled:
```
$ poetry install
Installing dependencies from lock file

Package operations: 0 installs, 0 updates, 8 removals

  • Removing cffi (1.14.3)
  • Removing cryptography (3.1.1)
  • Removing ecdsa (0.14.1)
  • Removing pyasn1 (0.4.8)
  • Removing pycparser (2.20)
  • Removing python-jose (3.2.0)
  • Removing rsa (4.6)
  • Removing six (1.15.0)
```

When adding the package in a different way, f.e. by adding the wheel by its direct URL https://pypi.thegraveyard.org/api/package/iksteen-aaa/iksteen_aaa-0.1.0-py3-none-any.whl, the uninstalling of the extra dependencies does NOT occur.

**Comment by abn** [comment]
@iksteen first off, thank you for the extremely useful bug report! :+1: 

Did a bit of digging; seems the issue is the way we parse the extras from the lock file. I have a hotfix for `1.1` and a proper fix for `1.2`. Will put up a PR soon.

**Comment by iksteen** [comment]
@abn That's awesome! Thanks for the quick triage!

**Comment by abn** [comment]
@iksteen Good bug reports help :) 

Can you please try the fix at #3229.

### Using pipx
```sh
pipx install --suffix=@3229 'poetry @ git+https://github.com/python-poetry/poetry.git@refs/pull/3229/head'
```

### Using a container (podman | docker)
```sh
podman run --rm -i --entrypoint bash python:3.8 <<EOF
set -xe
python -m pip install -q git+https://github.com/python-poetry/poetry.git@refs/pull/3229/head
install -d foobar
pushd foobar
curl -sLO https://gist.githubusercontent.com/iksteen/83a3d5b55a8274bfbb434d72676fea70/raw/c981a55f957f61766be0ff7c77e169017aaf8d56/pyproject.toml
poetry install --no-root
poetry add iksteen-aaa -E example
poetry install --no-root
EOF
```

**Comment by iksteen** [comment]
Works for me!
